### ISSUE TITLE
Add ability to update object label without pushing any datastream files

### ISSUE THREAD

**Issue by mjordan** [SFULibrary/islandora_datastream_crud#34]
Coming out of discussion at https://groups.google.com/forum/#!msg/islandora/JvJ3BbG6kro/7BD9kpQ3BAAJ, Datastream CRUD should provide a way to update object labels without having to push a DC or other datastream.

**Comment by mjordan** [comment]
I've got a basic implementation working. So far I have:

`drush islandora_datastream_crud_update_object_properties --user=admin --pid_file=/tmp/update_these_objects.txt` with options:

* `--owner`
* `--state`
* `--update_object_label_from_dc`
* `--update_object_label_from_mods`

However, I need some use cases to help us decide how to handle repeated title elements, and subelements. For the "update_object_label_from_X" options, should we only take the first title element, or should we concatenate all titles that are present (dc:title in the case of DC, mods:titleInfo/title in th case of MODS)? Do we care about MODS subTitle elements?

```xml
  <titleInfo>
    <title>I am the first title.</title>
    <subTitle/>
  </titleInfo>
  <titleInfo>
    <title>I am the second title.</title>
    <subTitle>A title for testing</subTitle>
  </titleInfo>
```

And what about `<titleInfo>`'s other subelements,  `<partNumber>`, `<partName>`, `<nonSort>`? I'd like to keep this simple and not get too caught up in handling elements that are rarely used.... does anyone have a real use case for including these subelements?

**Comment by petermacdonald** [comment]
The only subelement of <titleInfo> we use is in or MODS records for Islandora is <title>.  So we would be satisfied with making the label from just the <title>.

**Comment by markpbaggett** [comment]
This is tricky.  Currently, we handle updating fgsLabel with a little script post MODS replacement because of this exactly.  With form builder, we can control what goes to fgsLabel project to project, form to form, via the whole associate process, but I can say that our data dictionary / metadata mapping allows curators and metadata people / catalogers to do a lot of different things that makes this hard for us to automate.  

For instance, for some projects we have a /mods:titleInfo[supplied="yes"]/title and a /mods:titleInfo/title with no attributes. In this case, we map the former to the label because of it shows up in the template in the place that people think of as the preferred title. In the metadata display, we have to separate solr transforms so we can display and differentiate the two titles (Supplied title: and Title:).  

So, I doubt I've been any help here.

One thing that  may be helpful to add thought is that we make use of /titleInfo/nonSort for articles.  Not sure if others do this or not, but we have solr transform that concats nonsort and title and one that keeps them separate (this is probably silly, and I really don't know where we use which now that I think of it).

**Comment by wjlevay** [comment]
We do use `<nonSort>` for articles, but I'd be fine with excluding articles from the object labels. We also use `<subTitle>`, but again I'd be fine with excluding those from the labels.

We would also want to exclude any alternative titles from making their way into the label. Here's a real use case:

     <titleInfo>
          <title>Burned district San Francisco</title>
     </titleInfo>
     <titleInfo type="alternative">
          <title>San Francisco burned district</title>
     </titleInfo>

So we'd want a `mods:titleInfo[not(@type='alternative')]/mods:title` sort of thing.

**Comment by mjordan** [comment]
Nice set of use cases emerging! I think the best way to allow the user to determine what the source of the "label" is is to allow passing in an XPath expression as a drush option. It's ugly but it's no worse than passing regexes into `sed`.

So instead of the two drush options I named above:

* `--update_object_label_from_dc`
* `--update_object_label_from_mods`

we provide one, `--update_object_label`, that is accompanied by two additional options, `--source_dsid` and `--source_xpath`.  The default for `--source_xpath` is `/mods:titleInfo/title` (and the  default for `--source_dsid` is `MODS` (with possible values of `MODS` and `DC`).

@petermacdonald's use case is probably the most common (or at least is probably very common) and is taken care of by the default values for `--source_dsid` and `--source_xpath`:

`drush islandora_datastream_crud_update_object_properties --user=admin --pid_file=/tmp/update_these_objects.txt --update_object_label`

@wjlevay you'd be able to do:

`drush islandora_datastream_crud_update_object_properties --user=admin --pid_file=/tmp/update_these_objects.txt --update_object_label --source_xpath="mods:titleInfo[not(@type='alternative')]/mods:title"`

and @markpbaggett you'd be able to do:

`drush islandora_datastream_crud_update_object_properties --user=admin --pid_file=/tmp/update_these_objects.txt --update_object_label --source_xpath="/mods:titleInfo[supplied='yes']/title"`. Mark, I'm not sure how we'd handle concatenating the values from two source elements like you describe, unless you can come up with a single XPath that could do that.

Note that the double quotes around the XPath are required to let Drush interact with the shell properly.

How does this sound?

**Comment by wjlevay** [comment]
Sounds good to me. Looking forward to testing this out. Thanks, Mark!

**Comment by mjordan** [comment]
To test the label updating (and also owner and state updating), do the following:

1. Check out the issue-34 branch.
1. Load a few test objects.
1. Pull down the MODS datastreams for those objects (e.g., `drush islandora_datastream_crud_fetch_datastreams --user=admin --pid_file=/tmp/issue-34.pids --dsid=MODS --datastreams_directory=/tmp/issue34mods`)
1. Edit the title element of each MODS file.
1. Push the modified MODS datastream files back, e.g., `drush islandora_datastream_crud_push_datastreams --user=admin  --datastreams_source_directory=/tmp/issue34mods --datastreams_crud_log=/tmp/issue-134.log`).
Confirm that the object labels for the objects has not changed (they shouldn't have)
1. Run the following command (adjusting paths, etc. if necessary): `drush islandora_datastream_crud_update_object_properties --user=admin --pid_file=/tmp/issue-34.pids --source_dsid=MODS --source_xpath="//mods:titleInfo/mods:title" --update_object_label`
1. You can use more than one option with `islandora_datastream_crud_update_object_properties` at a time if  you want, e.g., ``drush islandora_datastream_crud_update_object_properties --user=admin --pid_file=/tmp/issue-34.pids --source_dsid=MODS --source_xpath="//mods:titleInfo/mods:title" --owner=foo --update_object_label``

If you prefer to test with DC titles, replace the titles within the DC datastreams of some test objects, push them up, and run: `drush islandora_datastream_crud_update_object_properties --user=admin --pid_file=/tmp/issue-34.pids --source_dsid=DC --source_xpath="//dc:title" --update_object_label`.

If multiple elements are returned by your XPath, the values that are used for the new labels are concatenated with an empty space. For example, a DC with two title elements like this:

```xml
  <dc:title>Winter Travel in the North, Atlin, B.C. 191 </dc:title>
  <dc:title>Second title</dc:title>
```
will produce a new label like "Winter Travel in the North, Atlin, B.C. 191 Second title" using the XPath --source_xpath="//dc:title". If you want to provide your own concatenation string, add the `--concat_string` option; for example, `--concat_string=" / "` will result in a label like this: "Winter Travel in the North, Atlin, B.C. 191 / Second title".

**Comment by mjordan** [comment]
I completely admit that concatenating multiple strings with ` / ` is not going to please everyone. I'm happy to add an additional option, something like `--implode_string` or `--concat_string` to allow you to pass in your own string.

**Comment by calhist** [comment]
Works as expected! Thanks again -- this is very helpful.

Note: Since I also have in my MODS a `<relatedItem>` like so:

     <relatedItem type="host">
          <titleInfo>
               <title>Los Angeles: 1932-33</title>
          </titleInfo>
          ...
     </relatedItem>

I changed my source_xpath to `/mods:mods/mods:titleInfo[not(@type='alternative')]/mods:title`

-Bill Levay @wjlevay

**Comment by mjordan** [comment]
@wjlevay Thanks very much for testing. I'll wait for a bit more feedback before merging into 7.x.

**Comment by mjordan** [comment]
@nate-rcl provided some thoughtful comments and questions in #18 that are relevant to this issue.

**Comment by markpbaggett** [comment]
@mjordan, haven't had a chance to test this yet, but thanks again for your work on this.  I expect your solution for my problem will work perfectly.

**Comment by mjordan** [comment]
Just pushed up to improvements:

* new drush option `--concat_string`, which allows you to specify what string to use to join concatenated / repeated strings into your object labels. Default is a single space. Thanks @nate-rcl for the feedback.
* you can now use any datastream, not just MODS and DC, as the source for labels.

**Comment by mjordan** [comment]
@markpbaggett no problem and good to hear you'll find it useful.

**Comment by mjordan** [comment]
I'd like to merge this into 7.x unless anyone has any objections.

**Comment by nate-rcl** [comment]
@mjordan +1

**Comment by mjordan** [comment]
Merged with 24de20787d5eef74785077b51ddee159830fdc9a. Closing.
